File: 1137-n-th-tribonacci-number/1137-n-th-tribonacci-number.py
class Solution:
    def __init__(self):
        self.memo = {}

    def tribonacci(self, n: int) -> int:
        if n in self.memo:
            return self.memo[n]

        match n:
            case 2:
                return 1
            case 1:
                return 1
            case 0:
                return 0
        
        result = self.tribonacci(n - 3) + self.tribonacci(n - 2) + self.tribonacci(n - 1)
        
        self.memo[n] = result
        return result

File: 1137-n-th-tribonacci-number/1137-n-th-tribonacci-number.cpp
class Solution {
public:
    unordered_map<int, int> memo;

    int tribonacci(int n) {
        if (memo.find(n) != memo.end())
            return memo[n];
        
        switch (n) {
            case 2:
                return 1;
            case 1:
                return 1;
            case 0:
                return 0;
        }

        int result = tribonacci(n - 3) + tribonacci(n - 2) + tribonacci(n - 1);
        this->memo[n] = result;
        
        return result;
    }
};